runtime/interpreter/interpreter_expression.go

```diff
@@ -275,6 +275,17 @@ func (interpreter *Interpreter) getReferenceValue(value Value, resultType sema.T
 	case NilValue, ReferenceValue:
 		// Reference to a nil, should return a nil.
 		// If the value is already a reference then return the same reference.
+		// However, we need to make sure that this reference is actually a subtype of the resultType,
+		// since the checker may not be aware that we are "short-circuiting" in this case
+
+		if !interpreter.IsSubTypeOfSemaType(value.StaticType(interpreter), resultType) {
+			panic(InvalidMemberReferenceError{
+				ExpectedType:  resultType,
+				ActualType:    interpreter.MustSemaTypeOfValue(value),
```

Comment:
Maybe re-use the value's static type when reporting the error, i.e. the error's actual type can just be a static-to-sema conversion of the static type used for the subtype check (`MustSemaTypeOfValue` basically also just callsv `value.StaticType(interpreter)`):

(Pseudo-code, not tested:)

```suggestion
		dynamicType := value.StaticType(interpreter)
		if !interpreter.IsSubTypeOfSemaType(dynamicType, resultType) {
			panic(InvalidMemberReferenceError{
				ExpectedType:  resultType,
				ActualType:    interpreter.MustConvertStaticToSemaType(dynamicType),
```